File: src/Historical/Fortran/adapter.f90
module temperature_adapter
  implicit none

  type :: legacy_fahrenheit_sensor
    integer :: fahrenheit = 86
  contains
    procedure :: read_fahrenheit
  end type legacy_fahrenheit_sensor

  type :: fahrenheit_sensor_adapter
    type(legacy_fahrenheit_sensor) :: adaptee
  contains
    procedure :: read_celsius
  end type fahrenheit_sensor_adapter

contains

  integer function read_fahrenheit(self)
    class(legacy_fahrenheit_sensor), intent(in) :: self
    read_fahrenheit = self%fahrenheit
  end function read_fahrenheit

  integer function read_celsius(self)
    class(fahrenheit_sensor_adapter), intent(in) :: self
    integer :: fahrenheit
    fahrenheit = self%adaptee%read_fahrenheit()
    read_celsius = ((fahrenheit - 32) * 5) / 9
  end function read_celsius

end module temperature_adapter

program adapter_demo
  use temperature_adapter
  implicit none

  type(legacy_fahrenheit_sensor) :: legacy
  type(fahrenheit_sensor_adapter) :: adapter

  adapter%adaptee = legacy
  write (*, '(A,I0,A)') 'legacy=', legacy%read_fahrenheit(), 'F'
  write (*, '(A,I0,A)') 'adapted=', adapter%read_celsius(), 'C'
end program adapter_demo

File: src/Historical/Pascal/adapter.pas
program Adapter;

{$mode objfpc}{$H+}

type
  TLegacyFahrenheitSensor = class
  public
    function ReadFahrenheit: Integer;
  end;

  TTemperatureReader = class abstract
  public
    function ReadCelsius: Integer; virtual; abstract;
  end;

  TFahrenheitSensorAdapter = class(TTemperatureReader)
  private
    FLegacy: TLegacyFahrenheitSensor;
  public
    constructor Create(ALegacy: TLegacyFahrenheitSensor);
    function ReadCelsius: Integer; override;
  end;

function TLegacyFahrenheitSensor.ReadFahrenheit: Integer;
begin
  Result := 86;
end;

constructor TFahrenheitSensorAdapter.Create(ALegacy: TLegacyFahrenheitSensor);
begin
  inherited Create;
  FLegacy := ALegacy;
end;

function TFahrenheitSensorAdapter.ReadCelsius: Integer;
begin
  Result := Round((FLegacy.ReadFahrenheit - 32) * 5.0 / 9.0);
end;

var
  Legacy: TLegacyFahrenheitSensor;
  Reader: TTemperatureReader;
begin
  Legacy := TLegacyFahrenheitSensor.Create;
  Reader := TFahrenheitSensorAdapter.Create(Legacy);
  try
    WriteLn('legacy=', Legacy.ReadFahrenheit, 'F');
    WriteLn('adapted=', Reader.ReadCelsius, 'C');
  finally
    Reader.Free;
    Legacy.Free;
  end;
end.
